From 86d46f14f7ff0cd89eff2246f8f5c18319b2ca56 Mon Sep 17 00:00:00 2001
From: Xiaopeng Han <hanxiaop8@outlook.com>
Date: Fri, 4 Mar 2022 22:52:52 +0800
Subject: Fix describe po not finding vs when protocol is https (#36713)

* fix

* add releasenote

* add integration test

* fix test

* unit test
---
 istioctl/cmd/describe.go                      |   3 +-
 istioctl/cmd/describe_test.go                 |  65 ++++++
 .../cmd/testdata/describe/http_config.json    | 199 ++++++++++++++++++
 .../cmd/testdata/describe/tls_config.json     | 199 ++++++++++++++++++
 releasenotes/notes/36713.yaml                 |   8 +
 5 files changed, 473 insertions(+), 1 deletion(-)
 create mode 100644 istioctl/cmd/testdata/describe/http_config.json
 create mode 100644 istioctl/cmd/testdata/describe/tls_config.json
 create mode 100644 releasenotes/notes/36713.yaml

diff --git a/istioctl/cmd/describe.go b/istioctl/cmd/describe.go
index e09270a33d..762900cbc5 100644
--- a/istioctl/cmd/describe.go
+++ b/istioctl/cmd/describe.go
@@ -667,7 +667,8 @@ func getIstioVirtualServicePathForSvcFromRoute(cd *configdump.Wrapper, svc v1.Se
 		if err != nil {
 			return "", err
 		}
-		if routeTyped.Name != sPort && !strings.HasPrefix(routeTyped.Name, "http.") {
+		if routeTyped.Name != sPort && !strings.HasPrefix(routeTyped.Name, "http.") &&
+			!strings.HasPrefix(routeTyped.Name, "https.") {
 			continue
 		}
 
diff --git a/istioctl/cmd/describe_test.go b/istioctl/cmd/describe_test.go
index 8b28136ef8..880f916142 100644
--- a/istioctl/cmd/describe_test.go
+++ b/istioctl/cmd/describe_test.go
@@ -20,11 +20,14 @@
 	"strings"
 	"testing"
 
+	"github.com/google/go-cmp/cmp"
 	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/kubernetes/fake"
 
+	"istio.io/istio/istioctl/pkg/util/configdump"
 	"istio.io/istio/pilot/test/util"
 )
 
@@ -181,3 +184,65 @@ func mockInterfaceFactoryGenerator(k8sConfigs []runtime.Object) func(kubeconfig
 
 	return outFactory
 }
+
+func TestGetIstioVirtualServicePathForSvcFromRoute(t *testing.T) {
+	tests := []struct {
+		name         string
+		inputConfig  string
+		inputService v1.Service
+		inputPort    int32
+		expected     string
+	}{
+		{
+			name:        "test tls config",
+			inputConfig: "testdata/describe/tls_config.json",
+			inputService: v1.Service{
+				TypeMeta: metav1.TypeMeta{},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "productpage",
+					Namespace: "default",
+				},
+				Spec:   v1.ServiceSpec{},
+				Status: v1.ServiceStatus{},
+			},
+			inputPort: int32(9080),
+			expected:  "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo",
+		},
+		{
+			name:        "test http config",
+			inputConfig: "testdata/describe/http_config.json",
+			inputService: v1.Service{
+				TypeMeta: metav1.TypeMeta{},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "productpage",
+					Namespace: "default",
+				},
+				Spec:   v1.ServiceSpec{},
+				Status: v1.ServiceStatus{},
+			},
+			inputPort: int32(9080),
+			expected:  "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ic, err := readFile(test.inputConfig)
+			if err != nil {
+				t.Fatalf("unable to open file in directory: %s", test.inputConfig)
+			}
+			cd := configdump.Wrapper{}
+			err = cd.UnmarshalJSON(ic)
+			if err != nil {
+				t.Fatal(err)
+			}
+			out, err := getIstioVirtualServicePathForSvcFromRoute(&cd, test.inputService, test.inputPort)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if diff := cmp.Diff(test.expected, out); diff != "" {
+				t.Fatalf("Diff:\n%s", diff)
+			}
+		})
+	}
+}
diff --git a/istioctl/cmd/testdata/describe/http_config.json b/istioctl/cmd/testdata/describe/http_config.json
new file mode 100644
index 0000000000..b245b4727d
--- /dev/null
+++ b/istioctl/cmd/testdata/describe/http_config.json
@@ -0,0 +1,199 @@
+{
+  "configs": [
+    {
+      "@type": "type.googleapis.com/envoy.admin.v3.RoutesConfigDump",
+      "dynamic_route_configs": [
+        {
+          "version_info": "2022-03-04T10:22:24Z/54",
+          "route_config": {
+            "@type": "type.googleapis.com/envoy.config.route.v3.RouteConfiguration",
+            "name": "http.8080",
+            "virtual_hosts": [
+              {
+                "name": "*:80",
+                "domains": [
+                  "*"
+                ],
+                "routes": [
+                  {
+                    "match": {
+                      "path": "/productpage",
+                      "case_sensitive": true
+                    },
+                    "route": {
+                      "cluster": "outbound|9080||productpage.default.svc.cluster.local",
+                      "timeout": "0s",
+                      "retry_policy": {
+                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
+                        "num_retries": 2,
+                        "retry_host_predicate": [
+                          {
+                            "name": "envoy.retry_host_predicates.previous_hosts"
+                          }
+                        ],
+                        "host_selection_retry_max_attempts": "5",
+                        "retriable_status_codes": [
+                          503
+                        ]
+                      },
+                      "max_grpc_timeout": "0s"
+                    },
+                    "metadata": {
+                      "filter_metadata": {
+                        "istio": {
+                          "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo"
+                        }
+                      }
+                    },
+                    "decorator": {
+                      "operation": "productpage.default.svc.cluster.local:9080/productpage"
+                    }
+                  },
+                  {
+                    "match": {
+                      "prefix": "/static",
+                      "case_sensitive": true
+                    },
+                    "route": {
+                      "cluster": "outbound|9080||productpage.default.svc.cluster.local",
+                      "timeout": "0s",
+                      "retry_policy": {
+                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
+                        "num_retries": 2,
+                        "retry_host_predicate": [
+                          {
+                            "name": "envoy.retry_host_predicates.previous_hosts"
+                          }
+                        ],
+                        "host_selection_retry_max_attempts": "5",
+                        "retriable_status_codes": [
+                          503
+                        ]
+                      },
+                      "max_grpc_timeout": "0s"
+                    },
+                    "metadata": {
+                      "filter_metadata": {
+                        "istio": {
+                          "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo"
+                        }
+                      }
+                    },
+                    "decorator": {
+                      "operation": "productpage.default.svc.cluster.local:9080/static*"
+                    }
+                  },
+                  {
+                    "match": {
+                      "path": "/login",
+                      "case_sensitive": true
+                    },
+                    "route": {
+                      "cluster": "outbound|9080||productpage.default.svc.cluster.local",
+                      "timeout": "0s",
+                      "retry_policy": {
+                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
+                        "num_retries": 2,
+                        "retry_host_predicate": [
+                          {
+                            "name": "envoy.retry_host_predicates.previous_hosts"
+                          }
+                        ],
+                        "host_selection_retry_max_attempts": "5",
+                        "retriable_status_codes": [
+                          503
+                        ]
+                      },
+                      "max_grpc_timeout": "0s"
+                    },
+                    "metadata": {
+                      "filter_metadata": {
+                        "istio": {
+                          "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo"
+                        }
+                      }
+                    },
+                    "decorator": {
+                      "operation": "productpage.default.svc.cluster.local:9080/login"
+                    }
+                  },
+                  {
+                    "match": {
+                      "path": "/logout",
+                      "case_sensitive": true
+                    },
+                    "route": {
+                      "cluster": "outbound|9080||productpage.default.svc.cluster.local",
+                      "timeout": "0s",
+                      "retry_policy": {
+                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
+                        "num_retries": 2,
+                        "retry_host_predicate": [
+                          {
+                            "name": "envoy.retry_host_predicates.previous_hosts"
+                          }
+                        ],
+                        "host_selection_retry_max_attempts": "5",
+                        "retriable_status_codes": [
+                          503
+                        ]
+                      },
+                      "max_grpc_timeout": "0s"
+                    },
+                    "metadata": {
+                      "filter_metadata": {
+                        "istio": {
+                          "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo"
+                        }
+                      }
+                    },
+                    "decorator": {
+                      "operation": "productpage.default.svc.cluster.local:9080/logout"
+                    }
+                  },
+                  {
+                    "match": {
+                      "prefix": "/api/v1/products",
+                      "case_sensitive": true
+                    },
+                    "route": {
+                      "cluster": "outbound|9080||productpage.default.svc.cluster.local",
+                      "timeout": "0s",
+                      "retry_policy": {
+                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
+                        "num_retries": 2,
+                        "retry_host_predicate": [
+                          {
+                            "name": "envoy.retry_host_predicates.previous_hosts"
+                          }
+                        ],
+                        "host_selection_retry_max_attempts": "5",
+                        "retriable_status_codes": [
+                          503
+                        ]
+                      },
+                      "max_grpc_timeout": "0s"
+                    },
+                    "metadata": {
+                      "filter_metadata": {
+                        "istio": {
+                          "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo"
+                        }
+                      }
+                    },
+                    "decorator": {
+                      "operation": "productpage.default.svc.cluster.local:9080/api/v1/products*"
+                    }
+                  }
+                ],
+                "include_request_attempt_count": true
+              }
+            ],
+            "validate_clusters": false
+          },
+          "last_updated": "2022-03-04T10:22:24.737Z"
+        }
+      ]
+    }
+  ]
+}
diff --git a/istioctl/cmd/testdata/describe/tls_config.json b/istioctl/cmd/testdata/describe/tls_config.json
new file mode 100644
index 0000000000..a31fc3085b
--- /dev/null
+++ b/istioctl/cmd/testdata/describe/tls_config.json
@@ -0,0 +1,199 @@
+{
+  "configs": [
+    {
+      "@type": "type.googleapis.com/envoy.admin.v3.RoutesConfigDump",
+      "dynamic_route_configs": [
+        {
+          "version_info": "2022-03-04T07:18:48Z/22",
+          "route_config": {
+            "@type": "type.googleapis.com/envoy.config.route.v3.RouteConfiguration",
+            "name": "https.443.https.bookinfo-gateway.default",
+            "virtual_hosts": [
+              {
+                "name": "*:443",
+                "domains": [
+                  "*"
+                ],
+                "routes": [
+                  {
+                    "match": {
+                      "path": "/productpage",
+                      "case_sensitive": true
+                    },
+                    "route": {
+                      "cluster": "outbound|9080||productpage.default.svc.cluster.local",
+                      "timeout": "0s",
+                      "retry_policy": {
+                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
+                        "num_retries": 2,
+                        "retry_host_predicate": [
+                          {
+                            "name": "envoy.retry_host_predicates.previous_hosts"
+                          }
+                        ],
+                        "host_selection_retry_max_attempts": "5",
+                        "retriable_status_codes": [
+                          503
+                        ]
+                      },
+                      "max_grpc_timeout": "0s"
+                    },
+                    "metadata": {
+                      "filter_metadata": {
+                        "istio": {
+                          "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo"
+                        }
+                      }
+                    },
+                    "decorator": {
+                      "operation": "productpage.default.svc.cluster.local:9080/productpage"
+                    }
+                  },
+                  {
+                    "match": {
+                      "prefix": "/static",
+                      "case_sensitive": true
+                    },
+                    "route": {
+                      "cluster": "outbound|9080||productpage.default.svc.cluster.local",
+                      "timeout": "0s",
+                      "retry_policy": {
+                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
+                        "num_retries": 2,
+                        "retry_host_predicate": [
+                          {
+                            "name": "envoy.retry_host_predicates.previous_hosts"
+                          }
+                        ],
+                        "host_selection_retry_max_attempts": "5",
+                        "retriable_status_codes": [
+                          503
+                        ]
+                      },
+                      "max_grpc_timeout": "0s"
+                    },
+                    "metadata": {
+                      "filter_metadata": {
+                        "istio": {
+                          "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo"
+                        }
+                      }
+                    },
+                    "decorator": {
+                      "operation": "productpage.default.svc.cluster.local:9080/static*"
+                    }
+                  },
+                  {
+                    "match": {
+                      "path": "/login",
+                      "case_sensitive": true
+                    },
+                    "route": {
+                      "cluster": "outbound|9080||productpage.default.svc.cluster.local",
+                      "timeout": "0s",
+                      "retry_policy": {
+                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
+                        "num_retries": 2,
+                        "retry_host_predicate": [
+                          {
+                            "name": "envoy.retry_host_predicates.previous_hosts"
+                          }
+                        ],
+                        "host_selection_retry_max_attempts": "5",
+                        "retriable_status_codes": [
+                          503
+                        ]
+                      },
+                      "max_grpc_timeout": "0s"
+                    },
+                    "metadata": {
+                      "filter_metadata": {
+                        "istio": {
+                          "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo"
+                        }
+                      }
+                    },
+                    "decorator": {
+                      "operation": "productpage.default.svc.cluster.local:9080/login"
+                    }
+                  },
+                  {
+                    "match": {
+                      "path": "/logout",
+                      "case_sensitive": true
+                    },
+                    "route": {
+                      "cluster": "outbound|9080||productpage.default.svc.cluster.local",
+                      "timeout": "0s",
+                      "retry_policy": {
+                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
+                        "num_retries": 2,
+                        "retry_host_predicate": [
+                          {
+                            "name": "envoy.retry_host_predicates.previous_hosts"
+                          }
+                        ],
+                        "host_selection_retry_max_attempts": "5",
+                        "retriable_status_codes": [
+                          503
+                        ]
+                      },
+                      "max_grpc_timeout": "0s"
+                    },
+                    "metadata": {
+                      "filter_metadata": {
+                        "istio": {
+                          "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo"
+                        }
+                      }
+                    },
+                    "decorator": {
+                      "operation": "productpage.default.svc.cluster.local:9080/logout"
+                    }
+                  },
+                  {
+                    "match": {
+                      "prefix": "/api/v1/products",
+                      "case_sensitive": true
+                    },
+                    "route": {
+                      "cluster": "outbound|9080||productpage.default.svc.cluster.local",
+                      "timeout": "0s",
+                      "retry_policy": {
+                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
+                        "num_retries": 2,
+                        "retry_host_predicate": [
+                          {
+                            "name": "envoy.retry_host_predicates.previous_hosts"
+                          }
+                        ],
+                        "host_selection_retry_max_attempts": "5",
+                        "retriable_status_codes": [
+                          503
+                        ]
+                      },
+                      "max_grpc_timeout": "0s"
+                    },
+                    "metadata": {
+                      "filter_metadata": {
+                        "istio": {
+                          "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bookinfo"
+                        }
+                      }
+                    },
+                    "decorator": {
+                      "operation": "productpage.default.svc.cluster.local:9080/api/v1/products*"
+                    }
+                  }
+                ],
+                "include_request_attempt_count": true
+              }
+            ],
+            "validate_clusters": false
+          },
+          "last_updated": "2022-03-04T07:18:49.007Z"
+        }
+      ]
+    }
+  ]
+}
diff --git a/releasenotes/notes/36713.yaml b/releasenotes/notes/36713.yaml
new file mode 100644
index 0000000000..36b1e36e4c
--- /dev/null
+++ b/releasenotes/notes/36713.yaml
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: istioctl
+issue:
+  - https://github.com/istio/istio/issues/35301
+releaseNotes:
+  - |
+    **Fixed** `describe pod` not showing the VirtualService info if the gateway is set to TLS ingress gateway.
-- 
2.35.3

